### PR TITLE
Setup es6 babel flow

### LINKED ISSUE #2: Create unit test with Jest
* create unit test with Jest

### PR DESCRIPTION
**Pull request by ayusharma**
Setup the following in repo
- Jest
- Flow
- Babel
- Eslint with prettier
- CircleCI
- Codecov

Fixes #2 #4 #5

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,49 @@
+version: 2
```
Comment (juanpicado) [inline .circleci/config.yml:1]: I'd use this version instead https://github.com/verdaccio/verdaccio-google-cloud/blob/master/.circleci/config.yml#L123 to publish from circleCI

File: package.json [hunk 2]
```diff
@@ -8,10 +9,28 @@
     "url": "git://github.com/dgaya/sinopia-memory.git"
   },
   "devDependencies": {
-    "mocha": "2.4.5"
-  },
-  "scripts": {
-    "test": "mocha ./tests"
+    "@commitlint/cli": "6.1.3",
+    "@commitlint/config-conventional": "6.1.3",
+    "@verdaccio/types": "^3.4.1",
```
Comment (juanpicado) [inline package.json]: It's better without `^`

File: package.json [hunk 3]
```diff
@@ -8,10 +9,28 @@
...
+    "babel-plugin-flow-runtime": "0.17.0",
+    "babel-preset-env": "1.7.0",
+    "babel-preset-flow": "6.23.0",
+    "codecov": "3.0.4",
+    "cross-env": "5.2.0",
+    "eslint": "5.2.0",
+    "eslint-config-prettier": "^2.9.0",
+    "eslint-plugin-prettier": "2.6.2",
+    "flow-bin": "0.76.0",
+    "flow-runtime": "0.17.0",
```
Comment (juanpicado) [inline package.json]: I think `flow-runtime` does not need it at all. The same applies to other projects. Can we do the test here?

File: src/index.js [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+ */
+export default class Memory {
+  // flow types
+  config: Object;
+  verdaccioConfig: Object;
+  users: Object;
+  max_users: number;
+  constructor(
+    config: { users: Object },
+    verdaccioConfig: { max_users: number }
```
Comment (juanpicado) [inline src/index.js]: Here you can create your own Flow object extending `Config`, check this https://github.com/verdaccio/verdaccio-google-cloud/blob/master/src/data-storage.js#L24

File: src/index.js [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+  ) {
+    this.config = config;
+    this.verdaccioConfig = verdaccioConfig;
+    this.users = config.users || {};
+    this.max_users = verdaccioConfig.max_users || Infinity;
+  }
+
+  authenticate(user: string, password: string, done: Function): Function {
```
Comment (juanpicado) [inline src/index.js]: `Function` should be `Callback`

File: src/index.js [hunk 6]
```diff
@@ -0,0 +1,106 @@
+// @flow
+/**
+ * Verdaccio auth memory
+ */
+export default class Memory {
```
Comment (juanpicado) [inline src/index.js]: implements https://github.com/verdaccio/flow-types/blob/master/src/types.js#L372

File: src/index.js [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+      error.status = 401;
+      return done(error);
+    }
+
+    // authentication succeeded!
+    // return all usergroups this user has access to;
+    return done(null, [user]);
+  }
+
+  adduser(user: Object, password: string, done: Function): Function {
```
Comment (juanpicado) [inline src/index.js]: user should be `RemoteUser`, check types

**Review by juanpicado — CHANGES_REQUESTED** [review]
🍰

File: src/index.js [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+
+  allow_access(user: RemoteUser, pkg: Object, done: Callback): Callback {
+    if (pkg.access.includes('$all') || pkg.access.includes('$anonymous')) {
+      return done(null, true);
+    }
+
+    const { name } = user;
+
+    if (!name) {
+      const error = Error('not allowed to access package');
```
Comment (juanpicado) [inline src/index.js:72]: For errors we should not use native `Error`, instead in verdaccio we use https://github.com/verdaccio/local-storage/blob/master/src/local-fs.js#L8 

For now we can let it like that, but, in some point we have to centralize this https://github.com/verdaccio/verdaccio/blob/master/src/lib/utils.js#L354 in a package

**Review by juanpicado — CHANGES_REQUESTED** [review]
Almos there ..

File: src/index.js [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+      error.status = 401;
+      return done(error);
+    }
+
+    // authentication succeeded!
+    // return all usergroups this user has access to;
+    return done(null, [user]);
+  }
+
+  adduser(user: Object, password: string, done: Callback): Callback {
```
Comment (juanpicado) [inline src/index.js]: `user: Object` .... `RemoteUser`

File: package.json [hunk 10]
```diff
@@ -8,10 +9,27 @@
...
+    "babel-preset-env": "1.7.0",
+    "babel-preset-flow": "6.23.0",
+    "codecov": "3.0.4",
+    "cross-env": "5.2.0",
+    "eslint": "5.2.0",
+    "eslint-config-prettier": "^2.9.0",
+    "eslint-plugin-prettier": "2.6.2",
+    "flow-bin": "0.76.0",
+    "husky": "^0.14.3",
+    "in-publish": "^2.0.0",
```
Comment (juanpicado) [inline package.json]: Remove all `^`

File: .npmignore [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+.npmignore
+.travis.yml
+*.tmp-*
+_storage/
+circle.yml
+travis.yml
+*.md
+coverage/
+jest.config.js
+jestEnvironment.js
```
Comment (juanpicado) [inline .npmignore:24]: You shouls exclude `src` here I guess.

**Review by juanpicado — CHANGES_REQUESTED** [review]
Sorry, minor changes.

File: .npmignore (hunk 11 above)
Comment (ayusharma) [reply .npmignore:24]: `src` is not present here.

**Review by juanpicado — APPROVED** [review]
👍

**Pull request merged**